src/script.js

```diff
@@ -0,0 +1,265 @@
+    })
+    .on('app-message', (e) => {
+      const { data } = e;
+
+      // If the isGameStarted property is not the one being updated,
+      // there's nothing further to do. Early out.
+      if (data.isGameStarted === undefined) return;
+      const isStarted = !!data.isGameStarted;
+
+      // If the game has just been started, enable the join integration
+      if (isStarted) {
+        updateIntegrations(callFrame, Integration.Player);
+        return;
+      }
+
+      // If the game has just been stopped, disable the join
+      // integration and enable the Kahoot host integration.
+      updateIntegrations(callFrame, Integration.Host);
+    })
+    .on('participant-joined', (e) => {
+      // If this player is a current Kahoot host, disable
+      // the local host integration and enable the player integration.
+      const isHost = !!e.participant.userData?.isKahootHost;
+      if (isHost) {
+        updateIntegrations(callFrame, Integration.Player);
+      }
+    });
+
+  return callFrame;
+}
+
+/**
+ * Leaves the Daily video call
+ * @param callFrame
+ */
+function leaveCall(callFrame) {
+  const welcomeScreen = document.getElementById('welcomeScreen');
+  const callContainer = document.getElementById('callContainer');
+  callContainer.classList.add(hideClassName);
+  welcomeScreen.classList.remove(hideClassName);
+  callFrame.destroy();
+}
+
+/**
+ * Updates the text of the Kahoot host button to represent current game state
+ * @param callFrame
+ * @param enableStart
+ */
+function updateKahootHostButton(callFrame, enableStart) {
+  const buttons = callFrame.customTrayButtons();
+  if (enableStart) {
+    buttons.kahootHost.label = 'Start Kahoot Game';
+    buttons.kahootHost.tooltip = 'Start Kahoot Game';
+  } else {
+    buttons.kahootHost.label = 'Exit Kahoot';
+    buttons.kahootHost.tooltip = 'Exit Kahoot';
+  }
+  callFrame.updateCustomTrayButtons(buttons);
+}
+
+/**
+ * Updates the local user's Kahoot host state and broadcasts
+ * that state as an app-message to all other participants.
+ * @param callFrame
+ * @param isStartingGame
+ */
+function updateGameHostState(callFrame, isStartingGame) {
+  callFrame.setUserData({
+    isKahootHost: isStartingGame,
+  });
+
+  callFrame.sendAppMessage({ isGameStarted: isStartingGame }, '*');
+}
+
+/**
+ * Gets the current Kahoot host, if any, and returns their Daily session ID
+ * @param callFrame
+ * @returns {undefined|string}
+ */
+function getCurrentKahootHost(callFrame) {
+  const allParticipants = callFrame.participants();
+  const participants = Object.values(allParticipants);
+  for (let i = 0; i < participants.length; i += 1) {
+    const p = participants[i];
+    if (p.userData?.isKahootHost) {
+      return p.session_id;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Enables the specified integrations (either Kahoot Host or Player),
+ * and disables the other one
+ * @param callFrame
+ * @param integration
+ */
+function updateIntegrations(callFrame, integration) {
```

Comment:
I think this is the main bit of refactoring from the original implementation. I realized when the HOST integration is active, the Player integration should always be inactive, and vice versa. In the past we had separate functions to toggle them and sometimes called those functions one after the other. 

I think having this single `updateIntegrations()` function is safer. Here, we can tell the function which integration should be active (host or player). The function will then always make sure the OTHER integration is disabled. 


I think this may also make it more intuitive to explain in writing, since all the integration toggling happens in this one place. WDYT @adamichelle ?

Reply:
Yeah that makes sense

Reply:
RE: my previous comment. I agree that it might be more intuitive in writing - tutorial wise - but it breaks some edge case functionality and may not be intuitive otherwise. For example, ensuring that the kahootHost integration is not visible when the kahootPlayer integration is started for the other users. 

I would make the case that the previous the implementation is however relatively easier to read since the functions are given descriptive names. They also do one thing and only one thing which how functions should typically behave. For that reason, I am not sure I am fully onboard with this change. I have tried to make changes to it to restore those broken functionality while keeping the overall idea but I have not been successful.

Reply:
Thanks, @adamichelle - you make really good points, especially about the previous implementation having more intuitive function naming.

My main concerns with the previous implementation were two-fold:

1. We had to "juggle" these two different function calls in various places, having to enable one integration and disable the other multiple times across different parts of the code.
2. We were calling `setCustomIntegrations()` (the Daily call frame instance method) several times one right after another just to turn one integration off and one on. 

My thinking with doing this all from a shared location was that it will reduce room for error (in that the caller has to only call _one_ function instead of two in the right order), and also reduce redundant API operations.

**BUT** you are totally right that having a big switch statement there can make it harder to read... I have another suggestion, which I have just submitted, and that is to pull out the modification of the integrations object into conveniently named functions. These can then be called from the _single_ `toggleIntegrations()` function (which still gives us _one_ place where this is all done.) The other difference is that the `setCustomIntegrations()` method is still then only called _once_ per update. 

Does this sound like a reasonable middle ground in terms of functionality vs readability?

Regarding the edge case functionality - it looks like that was just an error on my end, of not deleting the Kahoot custom button when the Kahoot Host integration is being disabled. Thanks for catching that! I've now added a fix. Are there any other edge cases you see that this change breaks? If I'm missing anything more I can take a look right away!

Phew, sorry about the wall of text :D Let me know if this alternative approach makes sense, or if it still doesn't feel quite right.

Reply:
Your approach resolved the issue and I think it is a reasonable middle ground. I am just adding the screenshot and modifying the class names and then maybe we can merge?